sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -218,7 +247,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       bucketSpec = getBucketSpec,
       options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    val destination = source match {
+      case "jdbc" => extraOptions.get(JDBCOptions.JDBC_TABLE_NAME)
+      case _ => extraOptions.get("path")
```

Comment:
Based on my understanding, the extra information we pass to QueryExecutionListener is like metadata. It is just for helping users understand the context. I still do not understand why we need to define a class/trait for it. This extra class/trait looks weird for this goal, unless you have some applications that are built on this class/trait.